usecases/ratelimiter/limiter.go

```diff
@@ -0,0 +1,59 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package ratelimiter
+
+import "sync"
+
+// Limiter is a thread-safe counter that can be used for rate-limiting requests
+type Limiter struct {
+	lock    sync.Mutex
+	max     int
+	current int
+}
+
+// New creates a [Limiter] with the specified maximum concurrent requests
+func New(maxRequests int) *Limiter {
+	return &Limiter{
+		max: maxRequests,
+	}
+}
+
+// If there is still room, TryInc, increases the counter and returns true. If
+// there are too many concurrent requests it does not increase the counter and
+// returns false
+func (l *Limiter) TryInc() bool {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	if l.max < 0 {
```

Comment:
That was the idea. So you can deactivate the rate limiter by passing in `-1`.

Reply:
In this case we could achieve the same behaviour by not using a rate limiter.

Reply:
But then we'd need an if branch in every place we use it. This way, it can act like a nice noop rate limiter, by simply deactivating it through config.

Reply:
in this case if `l.max < 0 `{ could be moved before` l.lock.Lock()` because l.max  is set in the constructor and is not changed afterwards and also not to incur the overhead of locking

Reply:
Correct, Andrzej spotted the same. The latest commit contains this change already.